src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs

```diff
@@ -361,6 +361,10 @@ public void Yield()
         /// </summary>
         public void Reacquire()
         {
+            // Release all cores on reacquire. The assumption here is that the task is done with CPU intensive work at this point and forgetting
+            // to release explicitly granted cores when reacquiring the node may lead to deadlocks.
```

Comment:
It could be a deadlock. The scheduler takes explicitly acquired cores into account when scheduling tasks. See the changes in `Scheduler.AtSchedulingLimit`: https://github.com/dotnet/msbuild/pull/5859/files#diff-a2f4cd20814bb6daa732376fd25d008cfdbabb8afad0eb9778a68e6165e43d5eR1389

Previous iterations of this PR used a "core counter" completely independent of task scheduling but my understanding is that it was a consequence of the implementation (semaphore), rather than done deliberately. We generally don't want to oversubscribe the machine much.